src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactoryHelpers.cs

```diff
@@ -0,0 +1,345 @@
+                    else
+                    {
+                        // Bind to ICollector<T>. 
+                        // This needs both a conversion from T to TMessage and an Sync/Async wrapper
+                        argumentBuilder = DynamicInvokeBuildICollectorArgument2(elementType, converterManager, buildFromAttribute, cloner);
+                    }
+                }
+            }
+
+            if (parameter.IsOut)
+            {
+                Type elementType = parameter.ParameterType.GetElementType();
+
+                if (elementType.IsArray)
+                {
+                    if (elementType == typeof(TMessage[]))
+                    {
+                        argumentBuilder = (attrResolved) =>
+                        {
+                            IAsyncCollector<TMessage> raw = buildFromAttribute(attrResolved);
+                            var invokeString = cloner.GetInvokeString(attrResolved);
+                            return new OutArrayValueProvider<TMessage>(raw, invokeString);
+                        };
+                    }
+                    else
+                    {
+                        // out TMessage[]
+                        var e2 = elementType.GetElementType();
+                        argumentBuilder = DynamicBuildOutArrayArgument2(e2, converterManager, buildFromAttribute, cloner);
+                    }
+                }
+                else
+                {
+                    // Single enqueue
+                    //    out TMessage
+                    if (elementType == typeof(TMessage))
+                    {
+                        argumentBuilder = (attrResolved) =>
+                        {
+                            IAsyncCollector<TMessage> raw = buildFromAttribute(attrResolved);
+                            var invokeString = cloner.GetInvokeString(attrResolved);
+                            return new OutValueProvider<TMessage>(raw, invokeString);
+                        };
+                    }
+                    else
+                    {
+                        // use JSon converter
+                        // out T
+                        argumentBuilder = DynamicInvokeBuildOutArgument2(elementType, converterManager, buildFromAttribute, cloner);
+                    }
+                }
+            }
+
+            if (argumentBuilder == null)
+            {
+                // Can't bind it. 
+                return null;
+            }
+
+            ParameterDescriptor param = new ParameterDescriptor
+            {
+                Name = parameter.Name,
+                DisplayHints = new ParameterDisplayHints
+                {
+                    Description = "output"
+                }
+            };
+
+            return new AsyncCollectorBinding<TAttribute, TMessage>(param, argumentBuilder, cloner);
+        }
+
+        private static Func<TAttribute, IValueProvider> DynamicBuildOutArrayArgument2<TAttribute, TMessage>(
```

Comment:
All the below methods have "2" suffixes as well - are you cleaning that up or something?